### ISSUE TITLE
Declaratively having multiple components of the same type?

### ISSUE THREAD

**Issue by ngokevin** [aframevr/aframe#1392]
**Description:**

How would we allow having multiple components of the same type, declaratively? If we should at all.

- Right now, you could create a custom parser and allow passing a complex string. Would be hard to code.
- You could do `setAttribute`, but you wouldn't be able to declaratively code that.
- Maybe a number suffix?

**Use cases:**

Set on event:

```html
<a-entity set-on-event-1="event: click; property: material.color; value: red"
               set-on-event-2="event: mouseenter; property: material.color: value: green; target: #enemy"></a-entity>
```

Randomizer:

```html
<a-entity random-1="property: material.color; min: 0 0 0; max: 0 0 0"
               random-2="property: geometry.primitive; oneOf: box circle sphere"
               random-3="property: material.roughness; min: 0; max: 0"></a-entity>
```

**Comment by dmarcos** [comment]
The first case is logic `when this happens do this` and should be in JS. You're cramming imperative logic in a declarative language. In a real world application the logic is usually more complicated than if `this event happens change this single component attribute and nothing else`. Having logic in both JS and your DOM will make things harder to understand and follow. 
I think I don't fully understand the second case.

**Comment by andreasplesch** [comment]
I believe the second case is about being able to randomize multiple properties in one invocation. Could parenting accomplish this ?
```
<a-entity random="property: material.color; min: 0 0 0; max: 1 0 0">
  <a-entity random="property: material.roughness; min: 0; max: 1">
    <a-entity  random="property: geometry.primitive; oneOf: box circle sphere"></a-entity>
  </a-entity>
</a-entity>
```
Would this also work for the first case ?

**Comment by ngokevin** [comment]
Declarative doesn't mean logicless. Many components should and do have logic. 

Declarative events is just an example. It is not expected by all means to handle all logic, but it is a common pattern. I just want it to fit better into the framework, even as a third-party component. Think CSS hover states, active states. When something happens (click, hover, collide, gaze), do something.

Everything is up front and there is no behavior hiding in a script tag somewhere. ECS makes it easy to encapsulate logic such that all I have to do is append to DOM to get things to execute:

```
<a-entity sound="on: click; src: scream.mp3"></a-entity>
<a-entity event-set-1="on: collide; property: explode"
               event-set-2="on: collide; property: material.color; color: red"></a-entity>
```

**Comment by ngokevin** [comment]
@andreasplesch That wouldn't work because then those would be modifying different entities. I'd want to be able to randomize any property value on a single entity.

**Comment by andreasplesch** [comment]
@ngokevin: Does this not work on rotation or position ? https://aframe.io/docs/guide/building-a-basic-scene.html#Interacting-with-the-Box
Hm, I thought children get their components from parents.
Could the component itself check if and what its parent entity randomizes and then add this to a list of randomized properties ?

**Comment by ngokevin** [comment]
We're thinking animation is a special case because it will be too complex to define as HTML attributes (e.g., sequencing, keyframing). Components are defined as HTML attributes.

For the randomizer component (again, it's just an example), it doesn't require any parent-child relationship. You could register a custom element to do it, but that breaks away from the ECS mold. We generally reserve parent-child relationship for the scene graph.

**Comment by andreasplesch** [comment]
I guess it is not ECS. There must be a recognized way then how ECS addresses this issue. Time to read.
Perhaps worth pointing out in the docs that parent-child only applies to rotation, position, scale(?), eg. that spatial transformations are an exception which take into account hierarchy.

**Comment by ngokevin** [comment]
For this issue specifically, the only problem we have is trying to define more complicated data structures in the DOM. With traditional ECS, you just use code so it's easy to pass around arrays of objects. 

rotation/position/scale aren't an exception. They apply to the entity they are attached to. It's just the way that local and nested coordinate systems affect each other, which is documented.

**Comment by dmarcos** [comment]
The declarative must not contain any logic to keep a good separations of concerns. Declarative: Describes what you want (HTML): A red cube, a cursor, a tiny elephant. Imperative: Describes how things will behave (JS). If you start adding logic to HTML you will have it scattered all over the place in JS and HTML making hard to follow the flow of your program. Writing the logic you propose above in JS is pretty trivial. You're optimizing for the trivial stuff and making the complicated stuff potentially more confusing and harder to understand.

**Comment by ngokevin** [comment]
This is unrelated to the PR, but I heavily disagree.

No, again, declarative does not mean no logic. It completely makes sense to declaratively define behavior and functionality. Components can define behavior and logic, and that's what makes them useful. Does that mean we should get rid of ```<a-entity sound="on: ...">```? That's logic. Should we get rid of controls? That's functionality.

It's not logic into HTML. The logic is encapsulated. It does not scatter it, but it encapsulates it into reusable pieces. If you put game/application logic into raw JS, then you no longer have control over its execution. Hell, it makes it easier to follow. Because it's defined right on the entity rather than in a bundle of ugly JS somewhere.

It's not about making it "trivial". It's about making it reusable.

**Comment by dmarcos** [comment]
the logic of controls is not expressed in the component? The component API just configures that logic/behavior? Sounds logic is to play a sound when an event is emitted. The API let's you define the name of that event, and that's all. It's not expressing any logic. `If this happens do this` That's logic and it's what you propose that it's different.

**Comment by dmarcos** [comment]
No, it's not easier to follow once things get complicated. It only works in toy examples. You will have to use JS in any non trivial application and then it will get confusing. Adding logic in declarative contexts has been tried many times and ends up being messy (angular I'm looking at you)

**Comment by ngokevin** [comment]
How is "play a sound when X event is emitted" any different from "change this property when X event is emitted"?  What if I just want to make something red when it's clicked on, collided, or hovered? That's pretty much an analog to CSS which declarative.

**Comment by dmarcos** [comment]
`play a sound when an event is emitted` is expresed in JS.

**Comment by ngokevin** [comment]
`change this property when X event is emitted` is also expressed in JS. It's encapsulated within the component.

**Comment by ngokevin** [comment]
Well. Ignore the declarative event example. There's controversy there. The other example just sets random property values, that doesn't contain logic. How could we define possibly multiple components of the same type?

The only way I can think of is create a DSL that's up to the dev, which might be hard to manage.

```<a-entity randomizer="property: material.color | property: position | property: scale; min: 1 1 1">```

**Comment by dmarcos** [comment]
CSS pseudo only changes the visual aspect of your UI. They don't interfere in how the application operates. Our components do much more than visual aspect.

**Comment by ngokevin** [comment]
Our components do much more than visual aspects, but shouldn't contain logic?

**Comment by ngokevin** [comment]
OK. I think I will go down the DSL path and custom parsers path.

I think the component arguments will solve themselves as we build more demos. Good thing we have permissionless innovation.
